---
title: "lecture_notes"
author: "Paslay"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Notes from 9/12/2024

## Genome alignment and assembly

### Alignment

- Why?
  - Reveal evolutionary relationships
  - Variant calling
  - Classify organisms
  - Gene expression
  - Epigenetic changes
  
- Pairwise alignment
Blast is a pairwise alignment tool. This maximizes the match and returns hits. We can do 1-1 or 1-database.

- Multiple sequence alignment
We can use ClustalOmega or align a number of sequences

- Aligning things to motifs

- What can we align?
DNA vs. DNA
RNA vs. RNA
DNA vs. RNA

- Terminology
Query: sequence to align
Reference: thing were aligning too
Homology: shared ancestry
Similarity, identity: mathematical way to detect homology
String: sequence

- Global vs. local alignment
global: each base has to align in some way betweeen sequences
local: picking regions to align (allowed to skip beginning and/or end of either sequence)

- We score alignments
Penalize mismatches and gaps
search possible alignments for the one that minimizes penalty (ex. mismatch = -1, deletion = -2, gap = -3)

M = match
X = mismatch
I = insertion (gap in target sequence)
D = deletion (gap in reference sequence)

**Hamming distance**: This does not take insertion or deletions into account but can quantify the number of changes from one sequence (A) to another sequence (B).

How do we find the best alignment?
This is based on the Smith-waterman algorithm (1981)
- One problem is that this algorithm may result in equally good alignments (aligns with the same score)
- It works well but is very computationally intensive. It requires n*m operations where `n` and `m` are sequence lengths.
- BLAST uses a combination of seeds and the SW-alignment method (+1 match, -2 mismatch, -0.5 for in/del)

E-value: number of hits one can expect to see by chance on a database this size. Small values are better matches.

BLAT is not BLAST
1) sequence vs genome (BLAT) instead of a sequence vs a database (BLAST)
2) only hits ...
3) Faster than BLAST, integrated into UCSC genome browser

Efficient alignment of millions of reads requires massive amounts of computation but the Burrows-Wheeler transform (BWT) can help:
- invented in 1994
- Rearranges and sorts seq elements
- Reduces computational cost
- Core component of most commonly used aligners

BWT -> FM-Index -> SW alignment

Tools:
- Bowtie2
- BWA-MEM
- Minimap2 (good for longer reads)

How to choose?
- Actively maintained
- Well documented
- Commonly used in your field
- Easy to install


Alignment output
- SAM: Sequence alignment map
- BAM: Binary Alignment MAP
  - For each read:
  - Coordinates in the reference
  - Sequences - carried over from fastq format
  - Alignment details - CIGAR sring
  - Quality information - Carried over from fastq format

### Flags

**Flags in the output file are very useful for selecting reads that aligned or that did not.**

```bash
samtools view -f 4 file.sam > unmapped.sam
samtools view -F 4 file.sam > mapped.sam
```

### CIGAR Strings

M - Alignment match or mismatch
= - Sequence match
X - sequence mismatch
I - Insertion to reference
D - Deletion from reference
S - soft clipping
H - Hard clipping
N - Skipped region

**Insert size**: Length of DNA fragment sequenced from both ends by paired-end sequencing

### BAM/SAM Visualization
The most common tool is the IGV (integrative genome viewer)


## Genome assembly
- Sequencing libraries consist of a million of small fragments of the genome
- How do we put these pieces back together?

Extract DNA -> fragment the DNA -> add adapters -> sequence by Illumina -> put the puzzle back together

**Coverage**: Typically 30-60x coverage for short read assembly is desired

1) Find overlapping regions and put these together (assemble by longest overlap)

2) accounting for other overlaps by creating a graph with the paths linking our fragments - Overlap Graph (classic approach) - Celera assembler (CA). This approach can work today with long reads from MinION or others


Now, we can simplify this by breaking the reads into sub-strings (kmers) of fixed lengths
- smaller than reads (k = 21-55)
- If small enough, we can find all unique kmers in sequencing dataset
- We still have the problem that there may be multiple overlaps or alternative paths


Another alternative is to create an overlap graph where we have kmers and nodes. The nodes would be visted once. This is called hometerian path


Eulerian path
- de Bruijn graph: kmers are the edges and the overlaps are the nodes
- visit every edge once
- Computationally tracable
- the graphs reduce redundancy
- Takes distinct kmers 
- Most assemblers use this method
- Limited by low sequencing depth and sequencing errors
- Repeats lead to forking paths which result in the **contigs**
- Limitation: repetitive sequences

Chromosome level assembly requires scaffolding
- link contigs with external reference, long reads, chromatin confirmation information (HiC)


### Chromosome confirmation - HiC

- Intrachromosomal contact porbability is on average much higher than interchromosomal.
- Interaction probability rapidly decays with increasing genomic distance
1. Cell or tissue sample
2. Crosslink chromatin
3. Fragment chromatin
4. repair and biotinylate ends
5. Ligate fragmented chromatin
6. Shear DNA and pull down DNA
7. Sequence with standard short-read seq


### Haploid assembly

- Having two alleles on an organism makes assembly challenging
- Haplotype phasing: requires long-reads

### Popular assemblers

SPAdes: 
- Designed to assemble megabase-sized genomes
- Multiple kmer de Bruijn graphs from short reads
- Can use long reads to resolve repeats

Hifiasm assembler:
- Build an overlap graph from HiFi reads
- Generate both haploid and diploid assemblies
- Can use very long reads to solve repeats
- Requires HiFi reads at a minimum


Unicycler:
- Build an overlap graph from long reads
- Polish the assembly
- Also has short-read options























  
- How?
- Limitations and Bias







### Assembly
- What and why?

